File: android/app/src/main/kotlin/com/kapybara/kapynotes/QuickAction.kt
package com.kapybara.kapynotes

import android.content.Context

/**
 * The three things a tap on a widget can ask the app for.
 *
 * Each is an intent action and nothing else. A widget knows no note, holds no
 * text, and reads no store: it names which of three doors the user came
 * through and lets the app decide what is behind it. That is what lets every
 * widget here be drawn once and never updated again, and what keeps a locked
 * phone from showing anybody's writing.
 */
enum class QuickAction(
    /** What the widget names on the intent that starts [MainActivity]. */
    val intentAction: String,
    /** The `LaunchIntent` Dart knows this by — lib/core/quick_capture.dart. */
    val launchName: String,
    val label: Int,
    val icon: Int,
) {
    /**
     * Named CONTINUE_WRITING, and staying that way. It is the action the
     * first version of this widget shipped with, and a widget already sitting
     * on somebody's Home Screen holds a `PendingIntent` built around it.
     */
    WRITE(
        "com.kapybara.kapynotes.CONTINUE_WRITING",
        "continueWriting",
        R.string.widget_action_write,
        R.drawable.ic_write_note,
    ),
    DICTATE(
        "com.kapybara.kapynotes.DICTATE",
        "dictate",
        R.string.widget_action_dictate,
        R.drawable.ic_dictate,
    ),
    CAPTURE(
        "com.kapybara.kapynotes.CAPTURE",
        "capture",
        R.string.widget_action_capture,
        R.drawable.ic_capture,
    ),
    ;

    companion object {
        /**
         * What a square widget does until somebody says otherwise: what the
         * widget did when Write was the only thing it could do.
         */
        val DEFAULT = WRITE

        fun ofIntentAction(action: String?): QuickAction? =
            entries.firstOrNull { it.intentAction == action }

        fun named(name: String?): QuickAction? = entries.firstOrNull { it.name == name }
    }
}

/**
 * Which action each placed square widget was configured to do.
 *
 * One small preferences file, written by the configuration screen and read by
 * the provider. It is keyed by widget id because two of the same widget on
 * the same Home Screen are two different questions — one may be Dictate and
 * the other Capture.
 */
object WidgetActions {
    private const val FILE = "com.kapybara.kapynotes.widget_actions"

    fun of(context: Context, appWidgetId: Int): QuickAction =
        QuickAction.named(prefs(context).getString(key(appWidgetId), null)) ?: QuickAction.DEFAULT

    fun put(context: Context, appWidgetId: Int, action: QuickAction) {
        prefs(context).edit().putString(key(appWidgetId), action.name).apply()
    }

    /** Called when a widget is removed, so the file does not grow forever. */
    fun forget(context: Context, appWidgetIds: IntArray) {
        val edit = prefs(context).edit()
        for (id in appWidgetIds) edit.remove(key(id))
        edit.apply()
    }

    private fun prefs(context: Context) =
        context.getSharedPreferences(FILE, Context.MODE_PRIVATE)

    private fun key(appWidgetId: Int) = "action.$appWidgetId"
}

File: ios/WriteNoteWidget/QuickAction.swift
import AppIntents
import SwiftUI
import UIKit

/// The three things a tap on a widget can ask the app for.
///
/// Each is a URL and nothing else. The widget knows no note, holds no text,
/// and reads no store: it says which of three doors the user came through and
/// lets the app decide what is behind it. That is what keeps every widget
/// here free to draw once and never refresh, and keeps a locked phone from
/// showing anybody's writing.
enum QuickAction: String, CaseIterable {
  case write
  case dictate
  case capture

  /// The word on the widget, and the name of the action everywhere else.
  var title: LocalizedStringResource {
    switch self {
    case .write: return "Write"
    case .dictate: return "Dictate"
    case .capture: return "Capture"
    }
  }

  /// What the widget gallery, the Lock Screen and Control Centre say it does.
  var summary: LocalizedStringResource {
    switch self {
    case .write: return "Carry on the note you were writing, ready to type."
    case .dictate: return "Carry on that note, and start talking into it."
    case .capture: return "Carry on that note, and add a picture to it."
    }
  }

  /// SF Symbols, which is what a Lock Screen tint and Control Centre both
  /// expect to be handed.
  var symbol: String {
    switch self {
    case .write: return "square.and.pencil"
    case .dictate: return "mic"
    case .capture: return "camera"
    }
  }

  /// What the app is opened on. The host is the whole message — see
  /// QuickCapture.swift in the Runner target, which turns it back into a
  /// `LaunchIntent` for Dart.
  var url: URL {
    URL(string: "kapynotes://\(rawValue)")!
  }
}

/// The paper and ink of the editor these open onto, so a tap looks like the
/// app unfolding rather than replacing something else. Matches
/// InstantCaptureApp in lib/ui/instant_capture.dart.
let paper = Color(
  uiColor: UIColor { traits in
    traits.userInterfaceStyle == .dark
      ? UIColor(red: 0.141, green: 0.125, blue: 0.094, alpha: 1)
      : UIColor(red: 0.969, green: 0.941, blue: 0.871, alpha: 1)
  })

let ink = Color(
  uiColor: UIColor { traits in
    traits.userInterfaceStyle == .dark
      ? UIColor(red: 0.929, green: 0.886, blue: 0.792, alpha: 1)
      : UIColor(red: 0.149, green: 0.212, blue: 0.290, alpha: 1)
  })

/// The same three, as something the widget editor can offer in a menu.
///
/// A separate type from [QuickAction] only because `AppEnum` drags display
/// representations and a localisation story in with it, and the widget's own
/// drawing code wants none of that.
enum QuickActionChoice: String, AppEnum {
  case write
  case dictate
  case capture

  static var typeDisplayRepresentation: TypeDisplayRepresentation = "Action"

  static var caseDisplayRepresentations: [QuickActionChoice: DisplayRepresentation] = [
    .write: DisplayRepresentation(
      title: "Write", subtitle: "Carry on the note you were writing.",
      image: .init(systemName: "square.and.pencil")),
    .dictate: DisplayRepresentation(
      title: "Dictate", subtitle: "Carry on that note, and start talking.",
      image: .init(systemName: "mic")),
    .capture: DisplayRepresentation(
      title: "Capture", subtitle: "Carry on that note, and add a picture.",
      image: .init(systemName: "camera")),
  ]

  var action: QuickAction { QuickAction(rawValue: rawValue) ?? .write }
}

/// What the user edits when they long-press the square widget and choose
/// **Edit Widget**.
///
/// The default is Write, which is what the widget did before it could be
/// configured at all — so a widget already on somebody's Home Screen keeps
/// doing exactly what they put it there to do.
struct SelectQuickActionIntent: WidgetConfigurationIntent {
  static var title: LocalizedStringResource = "Quick Action"
  static var description = IntentDescription("Choose what one tap does.")

  @Parameter(title: "Action", default: .write)
  var action: QuickActionChoice

  init() {}

  init(action: QuickActionChoice) {
    self.action = action
  }
}
